### ISSUE TITLE
Incorrectly tagged release : `2.13.3`

### ISSUE THREAD

**Issue by samveen** [xcat2/xcat-core#2886]
The tag for the `2.13.3` release is incorrectly created:
```
[samveen@buildhost2 xcat-core]$ git show 2.13.3
commit 7f9e0b57554c3a1975b83308935b7d183c632a9d
Merge: 7f09d26 b665e03
Author: zet809 <ertaozh@cn.ibm.com>
Date:   Thu Apr 13 13:43:38 2017 +0800

    Merge pull request #2863 from zet809/2.13
    
    Cherry-pick commits from master to 2.13 branch for 2.13.3 release(3)

```

Please compare to the tag for `2.13.2` release:
```
[samveen@buildhost2 xcat-core]$ git show 2.13.2
warning: refname '2.13.2' is ambiguous.
tag 2.13.2
Tagger: litingt@cn.ibm.com <litingt@cn.ibm.com>
Date:   Fri Feb 24 01:08:46 2017 -0500

xCAT 2.13.2 release

commit dab564dd7f23cb8bc46119f11d235de33c57802a
Merge: a113382 c1e0e43
Author: zet809 <ertaozh@cn.ibm.com>
Date:   Thu Feb 23 11:26:51 2017 +0800

    Merge pull request #2538 from xcat2/master
    
    merge master to 2.13 branch for 2.13.2 release.(2+)

```

This causes issues with builds that use `git describe`:
```
[samveen@buildhost2 xcat-core]$ git status
# HEAD detached at 2.13.3
nothing to commit, working directory clean
[samveen@buildhost2 xcat-core]$ git grep 'git describe'
buildcore.sh:        VER=`git describe`
makegenesisbuilderrpm:VER=`git describe`
[samveen@buildhost2 xcat-core]$ git describe
2.13.2-331-g7f9e0b5
```
expected output: `2.13.3`
Ref: http://stackoverflow.com/questions/4154485/git-describe-ignores-a-tag

**Comment by zet809** [comment]
Hi, @tingtli , will you pls help to take a look at this issue? Thx!

**Comment by immarvin** [comment]
hi @zet809 , would you please confirm with @tingtli  how she created the 2.13.3 tag?

thanks

**Comment by tingtli** [comment]
i am investigating this. Not sure why it is incorrect. I tag 2.13.3 same with older release.

**Comment by tingtli** [comment]
Is it because the commit is a cherry-pick?

**Comment by zet809** [comment]
Hi, @samveen , will you pull the update and check the "git describe" for 2.13 branch again? I had update the tag.

**Comment by samveen** [comment]
The problem is solved, however I expect something similar may happen for future tags created in branches other than master. That is a problem of my workflow, not an upstream issue. A brief description follows:
As the commit that was tagged as `2.13.3` was on the `2.13` branch, and not reachable from the `master` branch, `git fetch upstream` isn't fetching that tag for me (as I track just `upstream/master` in my fork). I had to force a fetch of all tags using `git fetch --tags`  to get it to work as expected. I'm modifying my workflow to bypass the issue in the future.

**Comment by whowutwut** [comment]
Joining this conversation late...  interesting. I think when I was experimenting with using git tags and commit numbers into the version of xCAT, I hit a similar problem because I was in the master and I couldn't get the right commit numbers since the tag in the branch...    I gave up at that time. 

How are other projects in git handling this? Do you know?  instead of creating release branches do they just tag a point in time on the master branch and that becomes the release copy? I know there should be release branches created so that testing can happen on the release branch and not affect normal development stream..

**Comment by samveen** [comment]
@whowutwut I found the same issue with the `2.13.4` tag. It is unreachable from the master branch. What I am wondering is that why is the `2.13.2` reachable from the master branch? How was it tagged differently, that it is reachable, but not the later `2.13.[34]` tags?

**Comment by zet809** [comment]
hi, @samveen , it is found that for the commit id that tag 2.13.3 point to in 2.13 branch, the same commit id occurs in master branch, so, git describe find that there is new tag created.

**Comment by whowutwut** [comment]
@samveen  This is still something I'm not fully understanding yet...

I think it has to do with the fact that we are not tagging on the master branch, so when we do `git describe` on master, it's showing the commits from 2.13.3 tag..    I guess we are expecting to see 2.13.4, but is it because there are no commits on 2.13 branch after the 2.13.4 tag is created so it doesn't reference from there?

**Comment by samveen** [comment]
@zet809 @whowutwut I just realised that `2.13.3` is reachable from `HEAD` on `master` because the `2.13` branch was merged back into `master`  via  commit `9000ae5d1eedc1c412d2a6f35bc0346ef33ad8af` which makes the tag `2.13.3` reachable from `HEAD`, while the 2.13.4 tag was created later.

I expect that the solution is to merge the 2.13 branch back to master _immediately_ **AFTER** the release tag is created.

**Comment by whowutwut** [comment]
@samveen  Interesting find....  

I think we do need to revisit the branching model we are following...  I have been wanting to use the `git describe` to have a "commit since last tag" number in the rpm name for xCAT but like this points out, we probably need to follow something like this: http://nvie.com/posts/a-successful-git-branching-model/ and release branches are tagged and then pushed into master, like what you found with that commit.  

But I'm not sure why that commit happened, since our current branching model should NOT be pushing branchs back into master.  So Prior to this last tag that we can see for 2.13, I think we only saw 2.13.0... and missed the .1 and .2 .  So something must have went wrong with our process that allowed that merge from 2.13 to master...  

So our process is the following. Development continues on `master` branch.. all pull requests are created there.  Then when we are ready to release, we would merge `master -> <release_branch>` in current cases. it's the 2.13 branch.  Then we would build against that branch to create Release candidates.   And we tag on the release branch, so we have this problem where we never see the tag from master branch.  

If we merged 2.13 to master right after we released, wouldn't we have some issues with any commits that have been merged to master branch since the merge of master -> release??

**Comment by samveen** [comment]
> If we merged 2.13 to master right after we released, wouldn't we have some issues with any commits that have been merged to master branch since the merge of master -> release??
@whowutwut This would normally apply for all release branches **except** the latest release branch(this case). 
Actually I expect that the `2.13` should have only been branched once work start on code for `2.14` starts. The latest release (highest value of X for all `2.X.Y` ) should always be tagged in master, getting branched to maintenance branches only on the next major release (`2.X+1.0`)

**Comment by daniceexi** [comment]
@whowutwut For your information that when @zet809 did the merge from master to 2.13 when preparing the release of 2.13.4, he made a mistake that merged the 2.13 to master to resolve certain conflict issue. This is why we saw the merge from 2.13 to master.
